src/rgw/rgw_lc.cc

```diff
@@ -1707,16 +1753,75 @@ int RGWLC::bucket_lc_process(string& shard_id, LCWorker* worker,
       return ret;
     }
 
-    op_env oenv(op, driver, worker, bucket.get(), ol);
-    LCOpRule orule(oenv);
-    orule.build(); // why can't ctor do it?
+    std::vector<LCOpRule> rules;
+    rules.reserve(active_ops.size());
+    for (auto* op : active_ops) {
+      op_env oenv(*op, driver, worker, bucket.get(), ol);
+      rules.emplace_back(oenv);
+      rules.back().build(); // why can't ctor do it?
+    }
+
     rgw_bucket_dir_entry* o{nullptr};
     for (auto offset = 0; ol.get_obj(this, yield, &o /* , fetch_barrier */); ++offset, ol.next()) {
-      orule.update();
-      workpool.spawn([&pf, dpp=this, orule, o=*o]
+      const auto obj = *o;
+
+      // Update all rules to capture current lister state before spawning
+      for (auto& rule : rules) {
+        rule.update();
+      }
+
+      // Spawn one coroutine per object to process all rules
+      workpool.spawn([&pf, dpp=this, rules_copy=rules, obj, bucket=bucket.get()]
                      (boost::asio::yield_context yield) mutable {
-          pf(dpp, yield, orule, o);
-        });
+        // Check if any rule needs tags so we only fetch once per object
+        bool any_rule_needs_tags = std::any_of(rules_copy.begin(), rules_copy.end(),
+          [](const LCOpRule& r) { return r.needs_tags(); });
+
+        boost::optional<RGWObjTags> cached_tags;
+        const RGWObjTags* cached_tags_ptr = nullptr;
+
+        if (any_rule_needs_tags && !obj.is_delete_marker()) {
+          bufferlist tags_bl;
+
+          rgw_obj_key obj_key = obj.key;
+          if (obj_key.instance.empty() && bucket->versioned() && !obj.is_current()) {
+            obj_key.instance = "null";
+          }
+
+          auto temp_obj = bucket->get_object(obj_key);
+          std::unique_ptr<rgw::sal::Object::ReadOp> rop = temp_obj->get_read_op();
+          int ret = rop->get_attr(dpp, RGW_ATTR_TAGS, tags_bl, yield);
+          if (ret == 0) {
+            try {
+              cached_tags.emplace();
+              auto iter = tags_bl.cbegin();
+              cached_tags->decode(iter);
+              cached_tags_ptr = &*cached_tags;
+            } catch (buffer::error& err) {
+              ldpp_dout(dpp, 5) << "ERROR: decode tags for " << obj.key << dendl;
+            }
+          }
+        }
+
+        for (auto& rule : rules_copy) {
+          if (rule.needs_tags() &&
+              !obj.is_delete_marker() &&
+              cached_tags_ptr &&
+              !has_all_tags(rule.get_op(), *cached_tags_ptr)) {
+            continue;
+          }
+
+          if (rule.needs_tags() &&
+              !obj.is_delete_marker() &&
+              !cached_tags_ptr) {
+            continue;
+          }
+
+          pf(dpp, yield, rule, const_cast<rgw_bucket_dir_entry&>(obj),
+             cached_tags_ptr);
+        }
```

Comment:
The per-object coroutine fetches tags into a local boost::optional<RGWObjTags> and immediately processes them, so nothing persists after that coroutine ends. Concurrency for those in-flight objects is bounded by the workpool limit. 

Worst case memory for tag prefetching is therefore “concurrency × tags per object”. With the S3 limits (10 tags, 128B key + 256B value) an RGWObjTags is only a few KB; at the default "rgw_lc_max_wp_worker <128>" that’s on the order of a megabyte or two. Even with a billion objects/tags total, we only ever hold the tags for the current in-flight set and release them immediately  after each object is processed.

Reply:
I think this resolves my objection, I agree one or a few MB is tolerable.